staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,601 @@
+
+	// information needed in the below loop to update a service
+	type serviceUpdateInfo struct {
+		// Service
+		service serviceKey
+
+		// Handler For the service which responds to /discovery/v1 requests
+		handler http.Handler
+
+		// ETag of the existing discovery information known about the service
+		etag string
+	}
+
+	var servicesToUpdate chan serviceUpdateInfo
+
+	// Collect all services which have no discovery document and then update them
+	// Done in two steps like this to avoid holding the lock while fetching the
+	// documents.
+	func() {
+		dm.servicesLock.Lock()
+		defer dm.servicesLock.Unlock()
+
+		servicesToUpdate = make(chan serviceUpdateInfo, len(dm.services))
+
+		// Close the buffered channel once we have finished populating if
+		// Note that even though it is closed it may still deliver any
+		// enqueued items to the webworkers.
+		defer close(servicesToUpdate)
+
+		for key, apiServiceInfo := range dm.services {
+			if !key.isLocalService() && len(apiServiceInfo.dependentAPIServices) == 0 {
+				// Purge unused non-local services from knowledge base
+				delete(dm.services, key)
+			} else if localOnly && !key.isLocalService() {
+				// If only local services are desired to be fetched, skip
+				// non-local
+				continue
+			} else if apiServiceInfo.fresh {
+				// Skip APIServices which are not marked dirty
+				continue
+			}
+
+			// Should not block. Buffered channel is as large as services
+			// slice
+			servicesToUpdate <- serviceUpdateInfo{
+				service: key,
+				handler: apiServiceInfo.handler,
+				etag:    apiServiceInfo.etag,
+			}
+		}
+	}()
+
+	// Download update discovery documents in parallel
+	type resultItem struct {
+		service   serviceKey
+		discovery *metav1.DiscoveryAPIGroupList
+		etag      string
+		error     error
+	}
+
+	waitGroup := sync.WaitGroup{}
+	results := make(chan resultItem, len(servicesToUpdate))
+
+	webworker := func() {
+		defer waitGroup.Done()
+
+		// Send a GET request to /discovery/v1 for each service that needs to
+		// be updated
+		// If channel is already closed, but has buffered items, will still only
+		// stop once all enqueued items have been processed
+		for updateInfo := range servicesToUpdate {
+			key := updateInfo.service
+			handler := updateInfo.handler
+			handler = handlerWithUser(handler, &user.DefaultInfo{Name: "system:kube-aggregator", Groups: []string{"system:masters"}})
+			handler = http.TimeoutHandler(handler, 5*time.Second, "request timed out")
+
+			req, err := http.NewRequest("GET", "/discovery/v1", nil)
+			if err != nil {
+				// NewRequest should not fail, but if it does for some reason,
+				// log it and continue
+				klog.Errorf("failed to create http.Request for /discovery/v1: %v", err)
+				continue
+			}
+			req.Header.Add("Accept", "application/json")
+
+			if updateInfo.etag != "" {
+				req.Header.Add("If-None-Match", updateInfo.etag)
+			}
+
+			writer := newInMemoryResponseWriter()
+			handler.ServeHTTP(writer, req)
```

Comment:
why are we wiring up an in-memory server acting as a proxy instead of issuing a normal "GET" for the discovery doc directly?

Reply:
This is issuing a normal "GET" for the discovery doc, similar to how the aggregator aggregates OpenAPI (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go#L57-L69)